File: api/models/schemas/inquiry.py
"""問い合わせ用Pydanticスキーマ.

APIリクエスト/レスポンスのシリアライゼーションとバリデーションを提供する。
"""

from datetime import datetime
from typing import Annotated, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator
from pydantic.functional_serializers import PlainSerializer

from models.enums.inquiry_status import InquiryStatus

# ISO 8601形式でシリアライズするdatetimeアノテーション
IsoDatetime = Annotated[
    datetime, PlainSerializer(lambda v: v.isoformat(), return_type=str)
]


class CreateInquiryRequest(BaseModel):
    """問い合わせ作成リクエスト.

    要件1.4: 問い合わせ内容を必須項目として検証する
    """

    user_id: str = Field(
        ...,
        min_length=1,
        max_length=50,
        pattern=r"^[a-zA-Z0-9_]+$",
        description="ユーザーID（英数字とアンダースコア、最大50文字）",
    )
    content: str = Field(
        ...,
        min_length=1,
        max_length=10000,
        description="問い合わせ内容（必須、最大10,000文字）",
    )
    source_system: str = Field(
        ...,
        min_length=1,
        max_length=50,
        description="送信元システム（例: manual, email, chat）",
    )

    @field_validator("content")
    @classmethod
    def validate_content_not_empty(cls, v: str) -> str:
        """contentが空白のみでないことを検証."""
        if not v.strip():
            raise ValueError("問い合わせ内容が空です")
        return v

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "user_id": "test_user",
                "content": "ログイン機能が欲しい",
                "source_system": "manual",
            }
        }
    )


class UpdateInquiryRequest(BaseModel):
    """問い合わせ更新リクエスト.

    要件2.8: 問い合わせ編集機能
    部分更新をサポートするため、すべてのフィールドはオプショナル。
    """

    content: Optional[str] = Field(
        None,
        min_length=1,
        max_length=10000,
        description="問い合わせ内容（最大10,000文字）",
    )
    source_system: Optional[str] = Field(
        None,
        min_length=1,
        max_length=50,
        description="送信元システム",
    )

    @field_validator("content")
    @classmethod
    def validate_content_not_empty(cls, v: Optional[str]) -> Optional[str]:
        """contentが提供された場合、空白のみでないことを検証."""
        if v is not None and not v.strip():
            raise ValueError("問い合わせ内容が空です")
        return v

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "content": "更新されたログイン機能の要望",
                "source_system": "email",
            }
        }
    )


class InquiryResponse(BaseModel):
    """問い合わせレスポンス.

    要件2.4: 各問い合わせについて ID、内容、ステータス、タイムスタンプを含む情報を返す
    タイムスタンプはISO 8601形式でシリアライズされる。
    """

    id: int = Field(..., description="問い合わせID")
    user_id: str = Field(..., description="ユーザーID")
    content: str = Field(..., description="問い合わせ内容")
    source_system: str = Field(..., description="送信元システム")
    timestamp: IsoDatetime = Field(..., description="問い合わせタイムスタンプ")
    status: InquiryStatus = Field(..., description="問い合わせステータス")
    created_at: IsoDatetime = Field(..., description="作成日時")
    updated_at: IsoDatetime = Field(..., description="更新日時")

    model_config = ConfigDict(
        from_attributes=True,  # ORMモデルからの変換を許可
        json_schema_extra={
            "example": {
                "id": 1,
                "user_id": "test_user",
                "content": "ログイン機能が欲しい",
                "source_system": "manual",
                "timestamp": "2025-12-27T00:00:00Z",
                "status": "received",
                "created_at": "2025-12-27T00:00:00Z",
                "updated_at": "2025-12-27T00:00:00Z",
            }
        },
    )


class ValidationErrorDetail(BaseModel):
    """バリデーションエラー詳細.

    個別のバリデーションエラー情報を表現する。
    """

    code: str = Field(..., description="エラーコード（GS-xxx形式）")
    message: str = Field(..., description="エラーメッセージ（日本語）")
    field: Optional[str] = Field(None, description="エラーが発生したフィールド名")

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "code": "GS-001",
                "message": "問い合わせ内容が空です",
                "field": "content",
            }
        }
    )


class ErrorResponse(BaseModel):
    """エラーレスポンス.

    要件1.3: エラーメッセージを報告者に表示する
    すべてのAPIエラーレスポンスで使用される統一フォーマット。
    """

    errors: list[ValidationErrorDetail] = Field(..., description="エラー情報のリスト")
    timestamp: IsoDatetime = Field(..., description="エラー発生日時")

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "errors": [
                    {
                        "code": "GS-001",
                        "message": "問い合わせ内容が空です",
                        "field": "content",
                    }
                ],
                "timestamp": "2025-12-27T00:00:00Z",
            }
        },
    )

File: web/src/types/inquiry.ts
/**
 * 問い合わせ管理機能 - TypeScript型定義
 *
 * バックエンドスキーマと完全に一致する型定義
 * 要件: 1.1, 1.2, 1.6, 2.4, 3.6, 3.7, 3.12
 */

/**
 * 問い合わせステータス列挙型
 */
export type InquiryStatus =
  | "received" // 受付済み
  | "processing" // AI処理中
  | "needs_clarification" // 明確化要求
  | "task_working" // タスク作業中
  | "completed" // 完了
  | "rejected" // 却下済み
  | "failed"; // 失敗

/**
 * 優先度列挙型
 */
export type Priority =
  | "low" // 低
  | "medium" // 中
  | "high" // 高
  | "urgent"; // 緊急

/**
 * ストーリーカテゴリ列挙型
 */
export type StoryCategory =
  | "development" // 開発
  | "testing" // テスト
  | "documentation" // ドキュメント
  | "research" // 調査
  | "maintenance" // メンテナンス
  | "custom"; // カスタム

/**
 * ステータス変更履歴エントリ
 */
export interface StatusHistoryEntry {
  from_status: InquiryStatus;
  to_status: InquiryStatus;
  changed_at: string; // ISO 8601形式
  changed_by?: string; // 変更者（将来実装）
}

/**
 * 却下情報
 */
export interface RejectionInfo {
  reason?: string; // 却下理由
  rejected_at: string; // 却下日時（ISO 8601）
  rejected_by?: string; // 却下者（将来実装）
}

/**
 * 問い合わせメタデータ構造
 */
export interface InquiryMetadata {
  // 却下情報（要件3.6-3.7）
  rejection?: RejectionInfo;

  // ステータス変更履歴（要件3.12）
  status_history?: StatusHistoryEntry[];

  // その他のメタデータ
  source?: string; // 送信元詳細情報
  tags?: string[]; // タグ（将来実装）
}

/**
 * 問い合わせレスポンス型
 * バックエンドAPIから返される問い合わせデータ
 */
export interface InquiryResponse {
  id: number;
  user_id: string;
  content: string;
  source_system: string;
  timestamp: string; // ISO 8601形式
  status: InquiryStatus;
  created_at: string; // ISO 8601形式
  updated_at: string; // ISO 8601形式
  inquiry_metadata?: InquiryMetadata;
}

/**
 * 問い合わせ作成リクエスト型
 */
export interface CreateInquiryRequest {
  user_id: string;
  content: string;
  source_system: string; // 送信元システム (例: "manual", "email", "chat")
}

/**
 * 問い合わせ更新リクエスト型
 */
export interface UpdateInquiryRequest {
  content?: string;
  source_system?: string;
}

/**
 * ページネーションメタデータ
 */
export interface PaginationMeta {
  page: number;
  limit: number;
  total: number;
  has_next: boolean;
}

/**
 * ページネーションレスポンス型
 */
export interface PaginatedResponse<T> {
  data: T[];
  meta: PaginationMeta;
  timestamp: string; // ISO 8601形式
}

/**
 * バリデーションエラー型
 */
export interface ValidationError {
  field: string;
  message: string;
  code: string; // GS-xxx形式
}

/**
 * エラーレスポンス型
 */
export interface ErrorResponse {
  errors: Array<{
    code: string; // GS-xxx形式
    message: string; // 日本語メッセージ
    field?: string; // バリデーションエラー時のフィールド名
  }>;
  timestamp: string; // ISO 8601形式
}

/**
 * 問い合わせ却下リクエスト型
 */
export interface RejectInquiryRequest {
  reason?: string; // 却下理由（任意）
}

/**
 * 問い合わせ一覧リクエストパラメータ型
 */
export interface ListInquiriesParams {
  page?: number; // デフォルト: 1
  limit?: number; // デフォルト: 20、範囲: 1-100
  status?: InquiryStatus | InquiryStatus[];
  user_id?: string;
  sort_by?: "created_at" | "updated_at";
  sort_order?: "asc" | "desc"; // デフォルト: 'desc'
}
